pkg/localstorev2/internal/stampindex/stamp.go

```diff
@@ -0,0 +1,163 @@
+		return nil, errStampItemMarshalNamespaceInvalid
+	}
+
+	var chImmutable byte = '0'
+	if i.chunkImmutable {
+		chImmutable = '1'
+	}
+
+	nsLen := len(i.namespace)
+	buf := make([]byte, 8+nsLen+swarm.HashSize+swarm.StampIndexSize+swarm.StampTimestampSize+1)
+
+	binary.LittleEndian.PutUint64(buf, uint64(len(i.namespace)))
+	copy(buf[8:], i.namespace)
+	copy(buf[8+nsLen:], i.batchID)
+	copy(buf[8+nsLen+swarm.HashSize:], i.batchIndex)
+	copy(buf[8+nsLen+swarm.HashSize+swarm.StampIndexSize:], i.batchTimestamp)
+	buf[8+nsLen+swarm.HashSize+swarm.StampIndexSize+swarm.StampTimestampSize+1] = chImmutable
+	return buf, nil
+}
+
+// Unmarshal implements the storage.Item interface.
+func (i *stampIndexItem) Unmarshal(bytes []byte) error {
+	if len(bytes) < 8 {
+		return errStampItemUnmarshalInvalidSize
+	}
+	nsLen := int(binary.LittleEndian.Uint64(bytes))
+	if len(bytes) != 8+nsLen+swarm.HashSize+swarm.StampIndexSize+swarm.StampTimestampSize+1 {
+		return errStampItemUnmarshalInvalidSize
+	}
+
+	ni := new(stampIndexItem)
+	ni.namespace = append(make([]byte, 0, nsLen), bytes[8:nsLen]...)
+	ni.batchID = append(make([]byte, 0, swarm.HashSize), bytes[8+nsLen:8+nsLen+swarm.HashSize]...)
+	ni.batchIndex = append(make([]byte, 0, swarm.StampIndexSize), bytes[8+nsLen+swarm.HashSize:8+nsLen+swarm.HashSize+swarm.StampIndexSize]...)
+	ni.batchTimestamp = append(make([]byte, 0, swarm.StampTimestampSize), bytes[8+nsLen+swarm.HashSize+swarm.StampIndexSize:8+nsLen+swarm.HashSize+swarm.StampIndexSize+swarm.StampTimestampSize]...)
+	switch bytes[8+nsLen+swarm.HashSize+swarm.StampIndexSize+swarm.StampTimestampSize] {
+	case '0':
+		ni.chunkImmutable = false
+	case '1':
+		ni.chunkImmutable = true
+	default:
+		return errStampItemUnmarshalChunkImmutableInvalid
+	}
+	*i = *ni
+	return nil
+}
+
+// String implements the fmt.Stringer interface.
+func (i stampIndexItem) String() string {
+	return path.Join(i.Namespace(), i.ID())
+}
+
+func LoadOrStore(s internal.Storage, namespace string, chunk swarm.Chunk) ([]byte, bool, error) {
+	timestamp, immutable, err := Load(s, namespace, chunk)
+	if errors.Is(err, storage.ErrNotFound) {
+		return chunk.Stamp().Timestamp(), chunk.Immutable(), Store(s, namespace, chunk)
+	}
+	return timestamp, immutable, nil
+}
+
+func Load(s internal.Storage, namespace string, chunk swarm.Chunk) ([]byte, bool, error) {
+	si := &stampIndexItem{
+		namespace:  []byte(namespace),
+		batchID:    chunk.Stamp().BatchID(),
+		batchIndex: chunk.Stamp().Index(),
+	}
+	err := s.Store().Get(si)
+	if err != nil {
+		return nil, false, fmt.Errorf("failed to get stampIndexItem %s: %w", si, err)
+	}
+	return si.batchTimestamp, si.chunkImmutable, nil
+}
+
+func Store(s internal.Storage, namespace string, chunk swarm.Chunk) error {
+	ns := []byte(namespace)
+	si := &stampIndexItem{
+		namespace:      ns,
+		batchID:        chunk.Stamp().BatchID(),
+		batchIndex:     chunk.Stamp().Index(),
+		batchTimestamp: chunk.Stamp().Timestamp(),
+		chunkImmutable: chunk.Immutable(),
```

Comment:
We need the chunk address stored for this index as well.

Reply:
Done.